Framework/MDAlgorithms/src/ConvertSpiceDataToRealSpace.cpp

```diff
@@ -616,22 +599,18 @@ IMDEventWorkspace_sptr ConvertSpiceDataToRealSpace::createDataMDWorkspace(
     detid_t detindex = 0;
 
     size_t nHist = thisWorkspace->getNumberHistograms();
+    const auto &specInfo = thisWorkspace->spectrumInfo();
     for (std::size_t i = 0; i < nHist; ++i) {
-      // For each spectrum/detector
-      Geometry::IDetector_const_sptr det = thisWorkspace->getDetector(i);
-      const MantidVec &vecsignal = thisWorkspace->readY(i);
-      const MantidVec &vecerror = thisWorkspace->readE(i);
+      const auto &vecsignal = thisWorkspace->y(i);
+      const auto &vecerror = thisWorkspace->e(i);
       float signal = static_cast<float>(vecsignal[0]);
       float error = static_cast<float>(vecerror[0]);
-      detid_t detid = det->getID() + detindex;
-      Kernel::V3D detPos = det->getPos();
-      double x = detPos.X();
-      double y = detPos.Y();
-      double z = detPos.Z();
+      detid_t detid = specInfo.detector(i).getID() + detindex;
+      const auto &detPos = specInfo.position(i);
       std::vector<Mantid::coord_t> data(3);
```

Comment:
As an optimization, consider using an array to avoid memory allocation by `std::vector`.